mem/idealmemcontroller/ctrlmiddleware.go

```diff
@@ -0,0 +1,127 @@
+package idealmemcontroller
+
+import "github.com/sarchlab/akita/v4/mem/mem"
+
+type ctrlMiddleware struct {
+	*Comp
+}
+
+func (m *ctrlMiddleware) Tick() (madeProgress bool) {
+	madeProgress = m.handleIncomingCommands() || madeProgress
+	madeProgress = m.handleStateUpdate() || madeProgress
+	return madeProgress
+}
+
+func (m *ctrlMiddleware) handleStateUpdate() (madeProgress bool) {
+	if m.state == "drain" {
+		madeProgress = m.handleDrainState() || madeProgress
+	}
+
+	return madeProgress
+}
+
+func (m *ctrlMiddleware) handleDrainState() bool {
+	if len(m.inflightBuffer) != 0 {
+		return false
+	}
+
+	rsp := m.currentCmd.GenerateRsp()
+
+	err := m.ctrlPort.Send(rsp)
+	if err != nil {
+		return false
+	}
+
+	m.state = "pause"
+	return true
+}
+
+func (m *ctrlMiddleware) handleIncomingCommands() (madeProgress bool) {
+	msg := m.ctrlPort.PeekIncoming()
+	if msg == nil {
+		return false
+	}
+
+	ctrlMsg := msg.(*mem.ControlMsg)
+
+	m.ctrlMsgMustBeValid(ctrlMsg)
+
+	madeProgress = m.handleEnable(ctrlMsg) || madeProgress
+	madeProgress = m.handlePause(ctrlMsg) || madeProgress
+	madeProgress = m.handleDrain(ctrlMsg) || madeProgress
+
+	return madeProgress
+}
+
+func (m *ctrlMiddleware) handleEnable(ctrlMsg *mem.ControlMsg) bool {
+	if ctrlMsg.Enable {
+		m.state = "enable"
+		rsp := ctrlMsg.GenerateRsp()
+		err := m.ctrlPort.Send(rsp)
+
+		if err != nil {
```

Comment:
It is recommended to have the `Send` and the error check in a single code block.